packages/shared/src/telemetry/clientCache.ts

```diff
@@ -0,0 +1,94 @@
+import type { TelemetryEventRaw } from './types';
+
+type TtlInMilliseconds = number;
+
+const DEFAULT_CACHE_TTL_MS = 86400000; // 24 hours
+
+/**
+ * Manages caching for telemetry events using the browser's localStorage to
+ * mitigate event flooding in frequently executed code paths.
+ */
+export class TelemetryClientCache {
+  #storageKey = 'clerk_telemetry';
+  #cacheTtl = DEFAULT_CACHE_TTL_MS;
+
+  cacheAndRetrieve(event: TelemetryEventRaw): boolean {
+    const now = Date.now();
+    const key = this.#generateKey(event);
+    const entry = this.#cache?.[key];
+
+    if (!entry) {
+      const updatedCache = {
+        ...this.#cache,
+        [key]: now,
+      };
+
+      localStorage.setItem(this.#storageKey, JSON.stringify(updatedCache));
+    }
+
+    const hasExpired = entry && now - entry > this.#cacheTtl;
```

Comment:
Realized this logic is incorrect. When an entry expires, we should update `clerk_telemetry` cache object and remove that key from it. 

When returning here, an event is cached when it isn't expired & it's entry exists.